src/daft-algebra/src/simplify/mod.rs

```diff
@@ -0,0 +1,151 @@
+mod boolean;
+mod null;
+mod numeric;
+
+use boolean::{simplify_binary_compare, simplify_boolean_expr};
+use common_error::DaftResult;
+use common_treenode::{Transformed, TreeNode};
+use daft_dsl::{lit, Expr, ExprRef};
+use daft_schema::schema::SchemaRef;
+use null::simplify_expr_with_null;
+use numeric::simplify_numeric_expr;
+
+/// Recursively simplify expression.
+pub fn simplify_expr(expr: ExprRef, schema: &SchemaRef) -> DaftResult<Transformed<ExprRef>> {
+    let simplify_fns = [
+        simplify_boolean_expr,
+        simplify_binary_compare,
+        simplify_expr_with_null,
+        simplify_numeric_expr,
+        simplify_misc_expr,
+    ];
+
+    expr.transform_up(|node| {
+        simplify_fns
+            .into_iter()
+            .try_fold(Transformed::no(node), |transformed, f| {
+                transformed.transform_data(|e| f(e, schema))
+            })
+    })
+}
+
+fn simplify_misc_expr(expr: ExprRef, schema: &SchemaRef) -> DaftResult<Transformed<ExprRef>> {
+    Ok(match expr.as_ref() {
+        // a BETWEEN low AND high --> a >= low AND e <= high
+        Expr::Between(between, low, high) => Transformed::yes(
+            between
+                .clone()
+                .lt_eq(high.clone())
+                .and(between.clone().gt_eq(low.clone())),
+        ),
+        // e IN () --> false
+        Expr::IsIn(_, list) if list.is_empty() => Transformed::yes(lit(false)),
+        // CAST(e AS dtype) -> e if e.dtype == dtype
+        Expr::Cast(e, dtype) if e.get_type(schema)? == *dtype => Transformed::yes(e.clone()),
```

Comment:
instead of erroring out, should we instead return `Transformed::no` if we can't `get_type`?

Reply:
I don't think there should be a reason why get_type should fail for a valid expression, right?